backends/gstreamer/datachannel.rs

```diff
@@ -0,0 +1,144 @@
+            channel
+                .emit("send-string", &[&Value::from(message)])
+                .map(|_| ())
+                .map_err(|e| WebRtcError::Backend(e.to_string()))
+        } else {
+            Err(WebRtcError::Backend("Dropped channel".to_owned()))
+        }
+    }
+}
+
+pub struct GStreamerWebRtcDataChannel {
+    channel: InnerDataChannel,
+    webrtc_thread: WebRtcThread,
+    callbacks: Arc<Mutex<WebRtcDataChannelCallbacks>>,
+}
+
+impl GStreamerWebRtcDataChannel {
+    pub fn new(
+        webrtc_thread: WebRtcThread,
+        webrtc: &gst::Element,
+        init: &WebRtcDataChannelInit,
+    ) -> Result<Self, String> {
+        let channel = webrtc
+            .emit(
+                "create-data-channel",
+                &[&init.label, &None::<gst::Structure>],
+            )
+            .map_err(|e| e.to_string())?;
+        let channel = channel
+            .expect("Invalid datachannel")
+            .get::<glib::Object>()
+            .map_err(|e| e.to_string())?
+            .expect("Invalid datachannel");
+
+        GStreamerWebRtcDataChannel::from(channel, webrtc_thread)
+    }
+
+    pub fn from(channel: glib::Object, webrtc_thread: WebRtcThread) -> Result<Self, String> {
+        let callbacks = Arc::new(Mutex::new(WebRtcDataChannelCallbacks::new()));
+
+        let callbacks_ = callbacks.clone();
+        channel
+            .connect("on-open", false, move |_| {
+                callbacks_.lock().unwrap().open();
+                None
+            })
+            .map_err(|e| e.to_string())?;
+
+        let callbacks_ = callbacks.clone();
+        channel
+            .connect("on-close", false, move |_| {
+                callbacks_.lock().unwrap().close();
+                None
+            })
+            .map_err(|e| e.to_string())?;
+
+        let callbacks_ = callbacks.clone();
+        channel
+            .connect("on-error", false, move |error| {
+                if let Some(error) = error[0]
+                    .get::<glib::error::Error>()
+                    .expect("Invalid GError")
+                {
+                    callbacks_
+                        .lock()
+                        .unwrap()
+                        .error(WebRtcError::Backend(error.to_string()));
+                }
+                None
+            })
+            .map_err(|e| e.to_string())?;
+
+        let callbacks_ = callbacks.clone();
+        channel
+            .connect("on-message-string", false, move |message| {
+                if let Some(message) = message[1]
+                    .get::<String>()
+                    .expect("Invalid data channel message")
+                {
+                    callbacks_.lock().unwrap().message(message);
+                }
+                None
+            })
+            .map_err(|e| e.to_string())?;
+
+        let channel = InnerDataChannel {
+            channel: glib::SendWeakRef::from(channel.downgrade()),
```

Comment:
This is going to explode sooner or later. The datachannel object is thread-safe, but until https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/issues/1168 is solved you'll have to put your own wrapper struct around it on which `Send`/`Sync` are implemented.